course/src/isometry.cc

```diff
@@ -142,5 +142,171 @@ Vector3 Vector3::cross(const Vector3& vector) const {
+        throw std::out_of_range("index out of range, Matrix3 only have three rows");
+    }
+    return (*this)[index];
+}
+
+const Vector3 Matrix3::col(uint32_t index) const {
+    if(index > 2) {
+        throw std::out_of_range("index out of range, Matrix3 only have three rows");
+    }
+    return Vector3(r1()[index], r2()[index], r3()[index]);
+}
+
+Vector3& Matrix3::operator[](const uint32_t index) {
+    switch(index) {
+        case 0:
+            return r1();
+        case 1:
+            return r2();
+        case 2:
+            return r3();
+        default:
+            throw std::out_of_range("Error. Invalid row index for Matrix3");
+    }
+}
+
+const Vector3& Matrix3::operator[](const uint32_t index) const {
+    switch(index) {
+        case 0:
+            return r1();
+        case 1:
+            return r2();
+        case 2:
+            return r3();
+        default:
+            throw std::out_of_range("Error. Invalid row index for Matrix3");
+    }
+}
+
+Matrix3 Matrix3::operator+(const Matrix3& matrix) const {
+    Matrix3 aux(*this);
+    aux += matrix;
+    return aux;
+}
+
+Matrix3 Matrix3::operator-(const Matrix3& matrix) const {
+    Matrix3 aux(*this);
+    aux -= matrix;
+    return aux;
+}
+
+Matrix3 Matrix3::operator*(const double& value) const {
+    Matrix3 aux(*this);
+    aux *= value;
+    return aux;
+}
+
+Matrix3 Matrix3::operator*(const Matrix3& matrix) const {
+    Matrix3 aux(*this);
+    aux *= matrix;
+    return aux;
+}
+
+Matrix3 Matrix3::operator/(const Matrix3& matrix) const {
+    Matrix3 aux(*this);
+    aux /= matrix;
+    return aux;
+}
+
+Matrix3& Matrix3::operator+=(const Matrix3& matrix) {
+    r1() += matrix.r1();
+    r2() += matrix.r2();
+    r3() += matrix.r3();
+    return *this;
+}
+
+Matrix3& Matrix3::operator-=(const Matrix3& matrix) {
+    r1() -= matrix.r1();
+    r2() -= matrix.r2();
+    r3() -= matrix.r3();
+    return *this;
+}
+
+Matrix3& Matrix3::operator*=(const double& value) {
+    r1() *= value;
+    r2() *= value;
+    r3() *= value;
+    return *this;
+}
+
+Matrix3& Matrix3::operator*=(const Matrix3& matrix) {
+    r1() *= matrix.r1();
+    r2() *= matrix.r2();
+    r3() *= matrix.r3();
+    return *this;
+}
+
+Matrix3& Matrix3::operator/=(const Matrix3& matrix) {
+    r1() /= matrix.r1();
+    r2() /= matrix.r2();
+    r3() /= matrix.r3();
+    return *this;
+}
+
+bool Matrix3::operator==(const Matrix3& matrix) const {
+    if ((r1() == matrix.r1()) && (r2() == matrix.r2()) && (r3() == matrix.r3())) {
+        return true;
+    }
+    return false;
+}
+
+bool Matrix3::operator!=(const Matrix3& matrix) const {
+    return !(*this == matrix);
+}
+
+Matrix3& Matrix3::operator=(const Matrix3& matrix){
+    r1() = matrix.r1();
+    r2() = matrix.r2();
+    r3() = matrix.r3();
+    return *this;
+}
+
+Matrix3& Matrix3::operator=(Matrix3&& matrix) {
+    r1() = std::move(matrix.r1());
+    r2() = std::move(matrix.r2());
+    r3() = std::move(matrix.r3());
+    return *this;
```

Comment:
This may be fine, but in other places you implement move my transferring the pointer to the buffer that holds the vectors.

As I said in other comment, using pointers in the case of Matrix is not necessary, but if you do better stick to that logic everywhere.